File: mathematica control system/LabVIEW/labview.lua
local com = require 'com'

local callmods = ParameterModifier(2)
callmods[0] = false
callmods[1] = true
callmods = luanet.make_array(ParameterModifier,{callmods})
local function vi_call(vi, args, values)
	local names
	if values == nil then
		names = {}
		values = {}
		for k,v in pairs(args) do
			names[#names + 1] = k
			values[#values + 1] = v
		end
	else
		names = args
	end
	names = luanet.make_array(Object, names)
	values = luanet.make_array(Object, values)
	args = luanet.make_array(Object, {names,values})
	local res,ok = com.invoke(vi, 'Call', args, callmods)
	return (res or args[1]),ok
end

__dir = debug.getinfo(1).source:gsub('/','\\'):match'@(.*)\\.*$' or io.popen'cd':read'*l'
VIDir = __dir..'\\.\\vi\\'

pcall(function()
require'lfs'
local dir = lfs.currentdir()
lfs.chdir(__dir..'\\.\\dll')
import 'LabVIEW.dll'
lfs.chdir(dir)
end)

import 'LabVIEW.dll'
AppCls = ApplicationClass()
_LoadVI = {}
function LoadVI(path, option, resv)
	if _LoadVI[path] then return _LoadVI[path] end
	if path:sub(-3) ~= '.vi' then path = VIDir..path..'.vi' end
	option = option or 16
	if resv == nil then resv = false end
	local vi = setmetatable({vi=AppCls:GetVIReference(path,'',resv,option)},{
	__index = function (self,key)
		return com.get(self.vi,key)
	end,
	__newindex = function (self,key,value)
		return com.set(self.vi,key,value)
	end,
	__call = function (self, ...)
		return vi_call(self.vi, ...)
	end
	})
	_LoadVI[path] = vi
	return vi
end

function CallVI(vi, args, values)
	if type(vi) == 'string' then
		vi = LoadVI(vi)
	end
	return vi(args, values)
end

function CurrentTimeString()
	return CallVI('CurrentTimeString', {'date/time string'}, {0})[0]
end

function CtrlGetValue(ctrl)
    return CallVI('CtrlGetValue', {'reference','variant'}, {ctrl,0})[1]
end

function CtrlSetValue(ctrl, value)
    local res,ok = CallVI('CtrlSetValue', {'reference','variant'}, {ctrl,value})
    return ok,res
end

function CtrlSignalValue(ctrl, value)
    local res,ok = CallVI('CtrlSignalValue', {'reference','variant'}, {ctrl,value})
    return ok,res
end

function GetAllCtrls(vi, prefix)
	prefix = prefix or ''
	if type(vi) == 'string' then
		vi = LoadVI(vi)
	end
	local r = CallVI('GetAllCtrls',{'VI','Prefix','Ctrls'},{vi.vi,prefix,0})[2]
	local dict = {}
	for i in luanet.each(r) do
		dict[i[0]] = i[1]
	end
	return dict
end

function Console(text)
	local vi = LoadVI('Console')
	vi{['print string'] = text}
	vi.FPWinOpen = true
end

function Queue(name, value)
	if value == nil then
        return CallVI('Queue',{'Name','Mode','Value In','Value Out'},{name,1,'',''})[3]
	else
        CallVI('Queue',{'Name','Mode','Value In','Value Out'},{name,0,val,''})
    end
end

if not pcall(debug.getlocal, 4, 1) then
Console(CurrentTimeString()..'\n')
ctrls = GetAllCtrls('Console')
table.foreach(ctrls,print)
--Console(CurrentTimeString()..'\n')
CtrlSetValue(ctrls.console,CurrentTimeString()..'\n')
LoadVI('Console').FPWinOpen = true
print(CtrlGetValue(ctrls.console))
end

File: mathematica control system/LabVIEW/LabVIEW.m
(* ::Package:: *)

If[$VersionNumber < 10,
System`Association[list:{__Rule}][key_]:=key/.list;
System`Association/:System`ReplaceAll[key_,System`Association[list:{__Rule}]]:=key/.list;
]

Needs["NETLink`"];
Off[NET::netexcptn];
Off[LoadNETAssembly::noload];
LoadNETAssembly@FileNameJoin@{DirectoryName@$InputFileName,"dll","LabVIEW.dll"};
On[NET::netexcptn];
Off[LoadNETAssembly::noload];

System`Each[lst_,expr_]:=(expr@@#)&/@lst;
System`Invert[assoc_Association]:=Association@Each[{Values@assoc,Keys@assoc}\[Transpose],Rule];

ClearAll[AppCls,LVLink,Inited];
VIDir=FileNameJoin[{DirectoryName@$InputFileName,"vi"}]<>"\\";

ClearAll[LoadVI,RunVI,AbortVI,MakeAndCastNETObject,CallVI,CtrlGetValue,CtrlSetValue,CtrlSignalValue,
GetAllCtrls,FindCtrl,Env,Queue,LinkLV,Init]

(*resv default to be True for LabVIEW 2011*)
LoadVI[viname_String,option_Integer:16,resv_Symbol:False]:=Module[{vipath,r},
vipath=If[StringLength@viname>3&&StringTake[viname,-3]==".vi"&&!StringFreeQ[viname,"\\"|"/"],viname,VIDir<>viname<>".vi"];
r=AppCls@GetVIReference[vipath,"",resv,option];
If[r=!=$Failed,LoadVI[viname]=r];
r
];
(* ReleaseNETObject/@DownValues[LoadVI][[1;;-2,2]] *)

RunVI[viname_String]:=Module[{vi=LoadVI[viname]},
If[vi@ExecState=!=LabVIEW`ExecStateEnum`eRunTopLevel,
vi@FPWinOpen=True;vi@Run[True]]];

AbortVI[viname_String]:=Module[{vi=LoadVI[viname]},
If[vi@ExecState===LabVIEW`ExecStateEnum`eRunTopLevel,vi@Abort[]]];

MakeAndCastNETObject[obj_Integer,type_]:=NETLink`CastNETObject[
NETLink`MakeNETObject[obj,"System.UInt32"],type];

MakeAndCastNETObject[obj_,type_]:=Module[{r=obj},
If[!NETLink`NETObjectQ[r],r=NETLink`MakeNETObject[r]];
If[r===$Failed,obj,
If[NETLink`InstanceOf[r,"System.__ComObject"],r,NETLink`CastNETObject[r,type]]]
];

CallVI[vi_?NETLink`NETObjectQ,paramNames_List,paramVals_List]:=Module[{param1,param2,r},
NETLink`BeginNETBlock[];
param1=NETLink`CastNETObject[NETLink`MakeNETObject[paramNames],"object"];
param2=NETLink`ReturnAsNETObject[MakeAndCastNETObject[MakeAndCastNETObject[#,"object"]&/@ paramVals,"object"]];
vi@Call[param1,param2];
r=NETLink`NETObjectToExpression[param2];
NETLink`EndNETBlock[];
r
];
CallVI[viname_String,paramNames_List,paramVals_List]:=CallVI[LoadVI[viname],paramNames,paramVals];

CtrlGetValue[ctrl_Integer]:=CallVI["CtrlGetValue",{"reference","variant"},{ctrl,""}][[-1]];

CtrlSetValue[ctrl_Integer,variant_]:=CallVI["CtrlSetValue",{"reference","variant"},{ctrl,variant}][[-1]];

CtrlSignalValue[ctrl_Integer,variant_]:=CallVI["CtrlSignalValue",{"reference","variant"},{ctrl,variant}][[-1]];

GetAllCtrls[vi_?NETLink`NETObjectQ,prefix_:""]:=Association[Rule@@#&/@(CallVI["GetAllCtrls",{"VI","Prefix","Ctrls"},{vi,prefix,""}][[-1]])];
GetAllCtrls[viname_String,prefix_:""]:=GetAllCtrls[LoadVI[viname],prefix];

FindCtrl[vi_?NETLink`NETObjectQ,name_String]:=CallVI["FindCtrl",{"VI","Name","Ctrl"},{vi,name,""}][[-1]];
FindCtrl[viname_String,name_String]:=FindCtrl[LoadVI[viname],name];

Env[name_String]:=Module[{value,names,values},
{value,names,values}=CallVI["Env",{"Name","Value Out","names","values"},{name,"","",""}][[2;;-1]];
If[Length[names]==0,value,
{value,Rule@@#&/@Transpose[{names,values}]}
]
];

Env[name_String,value_]:=(CallVI["Env",{"Name","Value In"},{name,value}];);

(*Peak[data_List, fitfunc_String] := 
  CallVI["Peak", {"2d Data", "Fitting Function", "Peak Center", 
     "Peak Width", "Peak Height"}, {Transpose[data], fitfunc, 0, 0, 
     0}][[-3 ;; -1]];*)

Queue[name_String,value_:""]:=CallVI["Queue",{"Name","Mode","Value In","Value Out"},{name,Boole[value===""],value,""}][[-1]];

Console[str_String]:=(CallVI["Console",{"print string"},{str}];);

LinkLV[name_String:"LabVIEW"]:=(
If[Head@LVLink===LinkObject,
Off[LinkObject::linkv,LinkObject::linkn];
MathLink`SetDaemon[LVLink,False];
MathLink`LinkRemoveInterruptMessageHandler[LVLink];
MathLink`RemoveSharingLink[LVLink];
LinkClose[LVLink];
On[LinkObject::linkv,LinkObject::linkn];
];
LVLink=LinkCreate[name,LinkProtocol->"SharedMemory"];
MathLink`AddSharingLink[LVLink,MathLink`AllowPreemptive->True,MathLink`ImmediateStart->True];
MathLink`LinkAddInterruptMessageHandler[LVLink];
MathLink`SetDaemon[LVLink,True];
);

InitLV[]:=(
Quiet[
NETLink`ReleaseNETObject[AppCls]; 
ClearAll[AppCls];
];
AppCls=NETLink`NETNew["LabVIEW.ApplicationClass"];
);

InitLV[];
